File: app/src/main/java/com/codingspezis/android/metalonly/player/stream/metadata/IcyStreamMeta.java
package com.codingspezis.android.metalonly.player.stream.metadata;

import android.os.Build;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by r on 17.09.14.
 * <p/>
 * CREDIT TO:
 * http://stackoverflow.com/questions/8970548/how-to-get-metadata-of-a-streaming-online-radio
 */
public class IcyStreamMeta {

    protected URL streamUrl;
    private Map<String, String> metadata;
    private boolean isError;
    private Map<String, String> data;

    /**
     * Arbitrary, high number to limit the StringBuffer's size that is used to extract metadata.
     * Hopefully this value is also small enough to prevent the OOM errors from <a href="https://github.com/ironjan/metal-only/issues/68>#68</a>.
     */
    private static final int MAX_STRING_BUFFER_SIZE = 4096;
    private static final int INITIAL_STRING_BUFFER_SIZE = 128;

    public IcyStreamMeta() {
        isError = false;
    }

    public static Map<String, String> parseMetadata(String metaString) {
        Map<String, String> metadata = new HashMap<>();
        String[] metaParts = metaString.split(";");
//        ÔStreamTitle='Maat - The Divine Slaughtering Of Mankin';
        Pattern p = Pattern.compile("(\\w+)='(.*)'$");
        Matcher m;
        for (String metaPart : metaParts) {
            m = p.matcher(metaPart);
            if (m.find()) {
                metadata.put((String) m.group(1), (String) m.group(2));
            }
        }

        return metadata;
    }

    /**
     * Get artist using stream's title
     *
     * @return String
     * @throws IOException
     */
    public String getArtist() throws IOException {
        data = getMetadata();

        if (!data.containsKey("StreamTitle"))
            return "";

        String streamTitle = data.get("StreamTitle");
        String title = streamTitle.substring(0, streamTitle.indexOf("-"));
        return title.trim();
    }

    /**
     * Get streamTitle
     *
     * @return String
     * @throws IOException
     */
    public String getStreamTitle() throws IOException {
        data = getMetadata();

        if (!data.containsKey("StreamTitle"))
            return "";

        return data.get("StreamTitle");
    }

    /**
     * Get title using stream's title
     *
     * @return String
     * @throws IOException
     */
    public String getTitle() throws IOException {
        data = getMetadata();

        if (!data.containsKey("StreamTitle"))
            return "";

        String streamTitle = data.get("StreamTitle");
        String artist = streamTitle.substring(streamTitle.indexOf("-") + 1);
        return artist.trim();
    }

    public Map<String, String> getMetadata() throws IOException {
        if (metadata == null) {
            refreshMeta();
        }

        return metadata;
    }

    synchronized public void refreshMeta() throws IOException {
        retreiveMetadata();
    }

    synchronized private void retreiveMetadata() throws IOException {
        URLConnection con;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT)
            con = new IcyURLConnection(streamUrl);
        else
            con = streamUrl.openConnection();
        con.setRequestProperty("Icy-MetaData", "1");
        con.setRequestProperty("Connection", "close");
        con.setRequestProperty("Accept", null);
        con.connect();
        int metaDataOffset = 0;
        Map<String, List<String>> headers = con.getHeaderFields();
        InputStream stream = con.getInputStream();

        if (headers.containsKey("icy-metaint")) {
            // Headers are sent via HTTP
            metaDataOffset = Integer.parseInt(headers.get("icy-metaint").get(0));
        } else {
            // Headers are sent within a stream
            StringBuilder strHeaders = new StringBuilder(INITIAL_STRING_BUFFER_SIZE);
            int charInt;
            while ((charInt = stream.read()) != -1) {
                if(strHeaders.length() >= MAX_STRING_BUFFER_SIZE){
                    strHeaders.delete(0, MAX_STRING_BUFFER_SIZE/2);
                }

                //noinspection NumericCastThatLosesPrecision loop condition makes sure that charInt is in [0, 255]
                strHeaders.append(((char) charInt));
                if (strHeaders.length() > 5 && (strHeaders.substring((strHeaders.length() - 4), strHeaders.length()).equals("\r\n\r\n"))) {
                    // getEndDate of headers
                    break;
                }
            }

            // Match headers to get metadata offset within a stream
            Pattern p = Pattern.compile("\\r\\n(icy-metaint):\\s*(.*)\\r\\n");
            Matcher m = p.matcher(strHeaders.toString());
            if (m.find()) {
                metaDataOffset = Integer.parseInt(m.group(2));
            }
        }

        // In case no data was sent
        if (metaDataOffset == 0) {
            isError = true;
            return;
        }

        // Read metadata
        int b;
        int count = 0;
        int metaDataLength = 4080; // 4080 is the max length
        boolean inData;
        StringBuilder metaData = new StringBuilder();
        // Stream position should be either at the beginning or right after headers
        while ((b = stream.read()) != -1) {
            count++;

            // Length of the metadata
            if (count == metaDataOffset + 1) {
                metaDataLength = b * 16;
            }

            inData = metaDataOffset < count  + 1 && count < (metaDataOffset + metaDataLength);
            if (inData) {
                if (b != 0) {
                    metaData.append((char) b);
                }
            }
            if (count > (metaDataOffset + metaDataLength)) {
                break;
            }
        }

        // Set the data
        metadata = IcyStreamMeta.parseMetadata(metaData.toString());

        // Close
        stream.close();

    }

    public boolean isError() {
        return isError;
    }

    public URL getStreamUrl() {
        return streamUrl;
    }

    public void setStreamUrl(URL streamUrl) {
        this.metadata = null;
        this.streamUrl = streamUrl;
        this.isError = false;
    }
}

File: core/src/main/kotlin/com/codingspezis/android/metalonly/player/stream/metadata/IcyStreamMeta.kt
package com.codingspezis.android.metalonly.player.stream.metadata

import java.io.IOException
import java.net.URL
import java.net.URLConnection
import java.util.HashMap
import java.util.regex.Matcher
import java.util.regex.Pattern

/**
 * Created by r on 17.09.14.
 *
 *
 * CREDIT TO:
 * http://stackoverflow.com/questions/8970548/how-to-get-metadata-of-a-streaming-online-radio
 */
class IcyStreamMeta(private val SDK_INT: Int, val streamUrl: URL) {

    private var metadata: Map<String, String>? = null
    var isError: Boolean = false
        private set
    private var data: Map<String, String>? = null

    /**
     * Get artist using stream's title

     * @return String
     * *
     * @throws IOException
     */
    val artist: String
        @Throws(IOException::class)
        get() {
            data = getMetadata()

            if (!data!!.containsKey("StreamTitle"))
                return ""

            val streamTitle = data!!["StreamTitle"]
            val title = streamTitle?.substring(0, streamTitle.indexOf("-"))
            return title?.trim { it <= ' ' } ?: "Unknown Artist"
        }

    /**
     * Get streamTitle

     * @return String
     * *
     * @throws IOException
     */
    val streamTitle: String
        @Throws(IOException::class)
        get() {
            data = getMetadata()

            val title = data!!["StreamTitle"]

            return title ?: ""
        }

    /**
     * Get title using stream's title

     * @return String
     * *
     * @throws IOException
     */
    val title: String
        @Throws(IOException::class)
        get() {
            data = getMetadata()

            if (!data!!.containsKey("StreamTitle")) {
                return ""
            }

            val streamTitle = data!!["StreamTitle"]
            val title = streamTitle?.substring(streamTitle.indexOf("-") + 1)
            return title?.trim { it <= ' ' } ?: "Unknown Title"
        }

    @Throws(IOException::class)
    fun getMetadata(): Map<String, String>? {
        if (metadata == null) {
            refreshMeta()
        }

        return metadata
    }

    @Synchronized @Throws(IOException::class)
    fun refreshMeta() {
        retreiveMetadata()
    }

    @Synchronized @Throws(IOException::class)
    private fun retreiveMetadata() {
        val con: URLConnection
        if (SDK_INT >= KITKAT_VERSION_CODE)
            con = IcyURLConnection(streamUrl)
        else
            con = streamUrl!!.openConnection()

        con.setRequestProperty("Icy-MetaData", "1")
        con.setRequestProperty("Connection", "close")

        con.connect()

        var metaDataOffset = 0
        val headers = con.headerFields
        val stream = con.getInputStream()

        if (headers.containsKey("icy-metaint")) {
            // Headers are sent via HTTP
            metaDataOffset = Integer.parseInt(headers["icy-metaint"]?.get(0))
        } else {
            // Headers are sent within a stream
            val strHeaders = StringBuilder(INITIAL_STRING_BUFFER_SIZE)
            do {
                if (strHeaders.length >= MAX_STRING_BUFFER_SIZE) {
                    strHeaders.delete(0, MAX_STRING_BUFFER_SIZE / 2)
                }

                val charInt: Int = stream.read()

                strHeaders.append(charInt.toChar())

                val hasMinLength = strHeaders.length > 5
                val lastFourChars = strHeaders.substring(strHeaders.length - 4, strHeaders.length)
                val haveNotFoundEndOfTitle = !(hasMinLength && lastFourChars == "\r\n\r\n")
            } while (charInt != -1 && haveNotFoundEndOfTitle)

            // Match headers to get metadata offset within a stream
            val p = Pattern.compile("\\r\\n(icy-metaint):\\s*(.*)\\r\\n")
            val m = p.matcher(strHeaders.toString())
            if (m.find()) {
                metaDataOffset = Integer.parseInt(m.group(2))
            }
        }

        // In case no data was sent
        if (metaDataOffset == 0) {
            isError = true
            return
        }

        // Read metadata
        var count = 0
        var metaDataLength = 4080 // 4080 is the max length
        val metaData = StringBuilder()
        // Stream position should be either at the beginning or right after headers
        do {
            val b: Int = stream.read()

            count++

            // Length of the metadata
            if (count == metaDataOffset + 1) {
                metaDataLength = b * 16
            }

           val inData = metaDataOffset < count + 1 && count < metaDataOffset + metaDataLength
            if (inData) {
                if (b != 0) {
                    metaData.append(b.toChar())
                }
            }
            if (count > metaDataOffset + metaDataLength) {
                break
            }

        } while (b != -1)

        metadata = parseMetadata(metaData.toString())

        stream.close()
    }

    companion object {

        /**
         * The value of Build.VERSION_CODES.KITKAT inlined to decouple this class from android.os.Build
         */
        private val KITKAT_VERSION_CODE = 19

        /**
         * Arbitrary, high number to limit the StringBuffer's size that is used to extract metadata.
         * Hopefully this value is also small enough to prevent the OOM errors from [](https://github.com/ironjan/metal-only/issues/68>#68</a>.
     ) */
        private val MAX_STRING_BUFFER_SIZE = 4096
        private val INITIAL_STRING_BUFFER_SIZE = 128

        private fun parseMetadata(metaString: String): Map<String, String> {
            val metadata = HashMap<String, String>()
            val metaParts = metaString.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val p = Pattern.compile("(\\w+)='(.*)'$")
            var m: Matcher
            for (metaPart in metaParts) {
                m = p.matcher(metaPart)
                if (m.find()) {
                    metadata.put(m.group(1) as String, m.group(2) as String)
                }
            }

            return metadata
        }
    }
}